helix-core/src/main/java/org/apache/helix/util/HelixUtil.java

```diff
@@ -210,19 +231,36 @@ public static String serializeByComma(List<String> objects) {
       String metadataStoreAddress, ClusterConfig clusterConfig,
       List<InstanceConfig> instanceConfigs, List<String> liveInstances,
       List<IdealState> idealStates, List<ResourceConfig> resourceConfigs) {
-    return getAssignmentForWagedFullAutoImpl(metadataStoreAddress, clusterConfig,
-        instanceConfigs, liveInstances, idealStates, resourceConfigs, false);
+    return getAssignmentForWagedFullAutoHelper(metadataStoreAddress, clusterConfig, instanceConfigs,
+        liveInstances, idealStates, resourceConfigs, false);
   }
 
-  /*
-   * If usePrefLists is set to true, the returned assignment is based on preference lists; if
-   * false, the returned assignment is based on partition state mapping, which may differ from
-   * preference lists.
-   */
-  private static Map<String, ResourceAssignment> getAssignmentForWagedFullAutoImpl(
+  private static Map<String, ResourceAssignment> getAssignmentForWagedFullAutoHelper(
+      ZkBucketDataAccessor zkBucketDataAccessor, HelixDataAccessor helixDataAccessor,
+      ClusterConfig clusterConfig, List<InstanceConfig> instanceConfigs, List<String> liveInstances,
+      List<IdealState> idealStates, List<ResourceConfig> resourceConfigs, boolean usePrefLists) {
+
+    // Copy the cluster config and make globalRebalance happen synchronously
+    // Otherwise, globalRebalance may not complete and this util might end up returning
+    // an empty assignment.
+    ClusterConfig globalSyncClusterConfig = new ClusterConfig(clusterConfig.getRecord());
+    globalSyncClusterConfig.setGlobalRebalanceAsyncMode(false);
+
+    // Create an instance of read-only WAGED rebalancer
+    ReadOnlyWagedRebalancer readOnlyWagedRebalancer =
+        new ReadOnlyWagedRebalancer(zkBucketDataAccessor, globalSyncClusterConfig.getClusterName(),
+            globalSyncClusterConfig.getGlobalRebalancePreference());
+
+    return getAssignmentForWagedFullAutoImpl(helixDataAccessor, readOnlyWagedRebalancer,
+        globalSyncClusterConfig, instanceConfigs, liveInstances, idealStates, resourceConfigs,
+        usePrefLists);
+  }
+
+  private static Map<String, ResourceAssignment> getAssignmentForWagedFullAutoHelper(
```

Comment:
This method is making the code here more complicated. I suggest not adding this one but making data accessor construction logic private methods and then call them repeatedly in the existing APIs.

Reply:
TFTR. If we make data accessor construction logic private, when adding a new REAT API, then the caller still need to pass a raw ZK address?